src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DefaultRazorDocumentMappingService.cs

```diff
@@ -18,6 +18,73 @@ namespace Microsoft.AspNetCore.Razor.LanguageServer
 {
     internal class DefaultRazorDocumentMappingService : RazorDocumentMappingService
     {
+        public override bool TryMapFromProjectedDocumentEdit(RazorCodeDocument codeDocument, TextEdit edit, out TextEdit newEdit)
+        {
+            newEdit = default;
+
+            var csharpSourceText = codeDocument.GetCSharpSourceText();
+            var range = edit.Range;
+            if (!IsRangeWithinDocument(range, csharpSourceText))
+            {
+                return false;
+            }
+
+            var startIndex = range.Start.GetAbsoluteIndex(csharpSourceText);
+            var endIndex = range.End.GetAbsoluteIndex(csharpSourceText);
+            var mappedStart = TryMapFromProjectedDocumentPosition(codeDocument, startIndex, out var hostDocumentStart, out _);
+            var mappedEnd = TryMapFromProjectedDocumentPosition(codeDocument, endIndex, out var hostDocumentEnd, out _);
+
+            // Ideal case, both start and end can be mapped so just return the edit
+            if (mappedStart && mappedEnd)
+            {
+                newEdit = new TextEdit()
+                {
+                    NewText = edit.NewText,
+                    Range = new Range(hostDocumentStart, hostDocumentEnd)
+                };
+                return true;
+            }
+
+            // For the first line of a code block the C# formatter will often return an edit that starts
+            // before our mapping, but ends within. In those cases, when the edit spans multiple lines
+            // we just take the last line and try to use that.
+            //
+            // eg in the C# document you might see:
+            //
+            //      protected override void BuildRenderTree(Microsoft.AspNetCore.Components.Rendering.RenderTreeBuilder __builder)
+            //      {
+            // #nullable restore
+            // #line 1 "/path/to/Document.component"
+            //    
+            //          var x = DateTime.Now;
+            //
+            // To indent the 'var x' line the formatter will return an edit that starts the line before,
+            // with a NewText of '\n            '. The start of that edit is outside our mapping, but we
+            // still want to know how to format the 'var x' line, so we have to break up the edit.
+            if (!mappedStart && mappedEnd && range.SpansMultipleLines())
+            {
+                // Construct a theoretical edit that is just for the last line of the edit that the C# formatter
+                // gave us, and see if we can map that.
+                // The +1 here skips the newline character that is found, but also protects from Substring throwing
+                // if there are no newlines (which should be impossible anyway)
+                var lastNewLine = edit.NewText.LastIndexOfAny(new char[] { '\n', '\r' }) + 1;
```

Comment:
If the edit itself is multiple lines of content will this accidentally throw away content? i.e. 3 lines would result in the first 2 being thrown away instead of just the first. Or is that not really possible in practice?

Reply:
For that to happen the compiler would have to change how many blank lines they emit after a line directive, which wouldn't make sense because then it would be pointing at the wrong line, but also I would rather that the tests fail if that change ever happens, than try to allow for it here.

Reply:
Update; I'm wrong, this totally can happen when there are multiple blank lines after the start of a code block, but the C# formatter will always format them to 0 indent, so there is nothing of value in them that we'd be throwing away. Will add a Debug.Assert for this though, which will also cover your other comment and ensuring the bit thrown away is only whitespace. It actually will only be newline characters.

Reply:
> Will add a Debug.Assert for this though, hich will also cover your other comment and ensuring the bit thrown away is only whitespace. It actually will only be newline characters.

Dopeness!